js/style/style.js

```diff
@@ -455,13 +455,15 @@ class Style extends Evented {
 
         if (this._removedLayers[id] && layer.source) {
             // If, in the current batch, we have already removed this layer
-            // and we are now re-adding it, then we need to clear (rather
-            // than just reload) the underyling source's tiles.
-            // Otherwise, tiles marked 'reloading' will have buffers that are
-            // set up for the _previous_ version of this layer, confusing
+            // and we are now re-adding it with a different `type`, then we
+            // need to clear (rather than just reload) the underyling source's
+            // tiles.  Otherwise, tiles marked 'reloading' will have buckets /
+            // buffers that are set up for the _previous_ version of this
+            // layer, causing, e.g.:
             // https://github.com/mapbox/mapbox-gl-js/issues/3633
+            const removed = this._removedLayers[id];
             delete this._removedLayers[id];
-            this._updatedSources[layer.source] = 'clear';
+            this._updatedSources[layer.source] = removed.type !== layer.type ? 'clear' : 'reload';
```

Comment:
> 1. I remove a layer, add a new layer with the same type and different property functions

My (probably incomplete) understanding of this is that, for a given style property:
 - If we're going from a constant value to DDS and we attempt to render a not-yet-reloaded tile, then, since the `Style` says DDS, the painter will look for a corresponding paint vertex buffer.  Not finding it, it'll fall back to the default color. (from conversation w/ @lucaswoj the other day.)
 - If we're going from DDS to a constant value and we attempt to render a not-yet-reloaded tile, then the painter will pull the value directly from the `Style`, so it doesn't matter that the tile isn't yet reloaded.
 - If we're just changing a DDS function, then when we attempt to render a not-yet-reloaded tile, we'll simply get the old DDS values from the previous dds buffer.

> 2. I remove a layer, add a new layer with a different type, remove that layer, add a new layer with the original type.

Ugh, good point -- although I think this actually points to a more general issue, which is that we do not handle adding and removing a layer within the same 'batch'